### PR TITLE
fix(spanner): retry or resume eos and rst_stream errors

### PR DESCRIPTION
**Pull request by thiagotnunes**
Retries or resumes the following errors when executing an execute streaming sql request:
- Unavailable error
- Internal EOS error
- Internal RST_STREAM error

This mirrors what is being done in some of the other clients: 
- Python (https://github.com/googleapis/python-spanner/pull/122/files)
- Java (https://github.com/googleapis/java-spanner/blob/master/google-cloud-spanner/src/main/java/com/google/cloud/spanner/PartitionedDmlTransaction.java#L98-L116)

closes: #7104

### CODE REVIEW

**Comment by thiagotnunes** [comment]
There is a build failure now, because the [following test](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb#L211-L213) expects an exception to be thrown (and for the sql query not to be retried). We need to verify if this is behaviour that we would like to keep or change.

**Comment by thiagotnunes** [comment]
Decided only to retry the internal errors (eos, rst_stream) and unavailable errors, but not aborted errors. This is the current behaviour of the Java library, although it retries aborted errors only for PDML. I think this might be a little tricky to implement in the ruby library, since `results` is used by both `execute_streaming_sql` and `pdmls`.

File: google-cloud-spanner/lib/google/cloud/spanner/results.rb
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/results.rb:233]: Why `exclamation mark` after the method? Do we modify any object in-place? Same for `retry_request! ` as well.

File: google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb
Comment (hengfengli) [inline google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb:187]: Why do we delete the `GRPC:: Unavailable`? `Aborted` is newly supported. We probably should keep both `Unavailable` and `Aborted`.

File: google-cloud-spanner/lib/google/cloud/spanner/results.rb
Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/results.rb:233]: Although it does not modify the object in-place it calls either `execute_streaming_sql` or `read` (in a nutshell performs side effects - not pure function). This is the reasoning behind this choice. Let me know if you'd like me to remove it.

**Review by hengfengli — APPROVED** [review]
LGTM. Please wait for @jiren's approval.

File: google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb
Comment (thiagotnunes) [reply google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb:187]: This test was failing, because we are now retrying `Unavailable` errors, such as in the Java / Python clients. Since this test was not specific to handling `Unavailable` errors, but re-throwing errors in the case of a failure, we changed the error to `Aborted`. Let me know if this makes sense.

File: google-cloud-spanner/lib/google/cloud/spanner/results.rb
Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/results.rb:233]: But that means even `client.execute_query` should also have `!` at the end. To me,  I would think adding `!` means that it modifies an object in-place.

Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/results.rb:233]: It's a private method. I am okay with any approach, but it would be better to have a consistent convention in client lib.

Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/results.rb:233]: No worries, I will remove the bangs. I agree with you that having a consistent convention is more important.

Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/results.rb]: Done.

**Comment by jiren** [comment]
@thiagotnunes PR looks good. LGMT after resolving review comment.

File: google-cloud-spanner/lib/google/cloud/spanner/results.rb
Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/results.rb]: Done.

**Review by jiren — APPROVED** [review]

**Comment by thiagotnunes** [comment]
@jiren @hengfengli it seems I found a bug in the implementation and I'd like to confirm with you. In the `results.row` we were retrying on `GRPC::BadStatus` errors (such as `Unavailable` or `Internal`). 

We do, however, use the class `Google::Cloud::Spanner::V1::Client` for executing the GRPC [methods](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb). 
All of the methods of this class have a conversion of `GRPC::BadStatus` errors to `Google::Cloud::Error`s (see [execute_streaming_sql](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb#L913) for instance).

That is why I changed the `rescue` clause of the `results` class to catch the wrapped errors instead.

File: google-cloud-spanner/lib/google/cloud/spanner/results.rb
Comment (thiagotnunes) [inline google-cloud-spanner/lib/google/cloud/spanner/results.rb:160]: Note that `GRPC::Core::CallError` is not a `GRPC::BadStatus` error, so it won't be wrapped.

**Comment by hengfengli** [comment]
@thiagotnunes I think this is caused by the recent generator migration. I wonder why no tests catch this problem.

File: google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb
Comment (hengfengli) [inline google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb:187]: Why does the old code work then? Where does it convert to `Google::Cloud::UnavailableError`?

Comment (hengfengli) [reply google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb:187]: The conversion happens here: https://github.com/googleapis/google-cloud-ruby/blob/e4bcb6ec8b8b525363b4add6f45516256aa33489/google-cloud-spanner/lib/google/cloud/spanner/results.rb#L149-L153

Comment (hengfengli) [reply google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb:187]: In your changes (google-cloud-spanner/lib/google/cloud/spanner/results.rb), Line 176 will do the same conversion.

Comment (hengfengli) [reply google-cloud-spanner/test/google/cloud/spanner/client/read_resume_buffer_bound_test.rb:187]: But you're right. We can assume that all `GRPC::BadStatus` are converted to `Google::Cloud::Error` in the gapic layer.

**Comment by hengfengli** [comment]
This means that we probably need to fix all `rescue` code trying to catch an `GRPC::BadStatus` from gapic layer. Because now we can only get `Google::Cloud::Error`.

File: google-cloud-spanner/lib/google/cloud/spanner/results.rb
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/results.rb:179]: This code becomes redundant, right?

Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/results.rb:179]: Yes, it would.

**Comment by thiagotnunes** [comment]
@hengfengli I am re-running the PDML long transaction and seeing if it works with the new code. Will keep you posted.

**Comment by jiren** [comment]
> @thiagotnunes I think this is caused by the recent generator migration. I wonder why no tests catch this problem.

Yes, this is a miss. Not much tests are available for retry and resume. So I think we should add it.

File: google-cloud-spanner/lib/google/cloud/spanner/results.rb
Comment (jiren) [inline google-cloud-spanner/lib/google/cloud/spanner/results.rb:180]: @thiagotnunes  `rescue GRPC::BadStatus`  need to remove. Generator always going to throw `Google::Cloud::Error` so no need to capture again and cast.

**Review by jiren — CHANGES_REQUESTED** [review]

**Comment by thiagotnunes** [comment]
@jiren @hengfengli so, I don't think that the problem is as simple as I originally thought. The fact is that the Generator client (`Google::Cloud::Spanner::V1::Client`) is returning an `Enumerable` for the `streaming` requests (see [here](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb#L875)). 

This means that even though we `rescue` on `GRPC::BadStatus` and transform it to `Google::Cloud::Error` when invoking the streaming method in `Google::Cloud::Spanner::V1::Client` (as in [here](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb#L913)), we do not do the same when we call `Enumerable.next` (as in [here](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner/lib/google/cloud/spanner/results.rb#L116)).

Because of this, we can get a either a `Google::Cloud:Error` or a `GRPC:Error`. This is observed when I run the PDML and only retry `Google::Cloud::Error`s or only retry `GRPC::Error`s, I still sometimes get an error (either `Google::Cloud::InternalError` or `GRPC::Internal`) which bubbles up.

So we have a couple of choices here:
1. We can retry both errors in the `Results.row` ([here](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner/lib/google/cloud/spanner/results.rb#L146)). This is the simplest, less risky change, but I think this will add tech debt.
2. We can change `Results.from_enum` to also retry on this [method](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner/lib/google/cloud/spanner/results.rb#L230), which is the first invocation of RPC. This way we would only have to retry `GRPC::Error`s.
3. We can change the `Results.from_enum` ([here](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner/lib/google/cloud/spanner/results.rb#L229)), `@enum` field to be wrapped in a code block that `rescues` `GRPC::BadStatus` and transforms them to `Google::Cloud::Error`. This way we only need to handle `Google::Cloud::Error`. 
4. I don't know if this is possible, but could we change `Google::Cloud::Spanner::V1::Client` to return a `Enumerable` that always transforms the `GRPC::BadStatus` to `Google::Cloud::Error` ([here](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb#L908) and [here](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb#L1236))? With this, we would only need to `rescue` `Google::Cloud::Error`s.
5. Any other choices you can think of?

Personally I would vote for **option 4**, since I think that the intention was to encapsulate GRPC specific errors on the generated client abstraction and only expose `Google::Cloud::Error`s to the users. However, I would like to hear your opinion before implementing anything.

**Comment by hengfengli** [comment]
@thiagotnunes Nice find! My suggestion is that: 

* For this change, we deal with both kind of errors. 
* We file an issue with the ruby gapic generator team (@dazuma). https://github.com/googleapis/gapic-generator-ruby

**Comment by thiagotnunes** [comment]
@hengfengli thanks! I agree with your suggestion. I will make the PR changes here to reflect what we discussed and create an issue with the generator team.

File: google-cloud-spanner/lib/google/cloud/spanner/results.rb
Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/results.rb:191]: We will need to keep this for the time being (refer to the discussion below).

**Comment by thiagotnunes** [comment]
@jiren @hengfengli updated the code to handle both google cloud errors and grpc errors.

**Comment by hengfengli** [comment]
LGTM.

**Comment by thiagotnunes** [comment]
@hengfengli thanks, created the following issue with the ruby gapic generator team: https://github.com/googleapis/gapic-generator-ruby/issues/509

**Comment by thiagotnunes** [comment]
@jiren could you re-review? The change that you requested is no longer valid, since we need to `rescue` both at the moment. Once the gapic generator team has implemented the feature linked above, we will be able to only `rescue` `Google::Cloud::Error`s.

**Review by jiren — APPROVED** [review]

**Pull request merged**
